### ISSUE TITLE
Parse Installation

### ISSUE THREAD

**Issue by novemTeam** [parse-community/Parse-Swift#93]
I have an issue where the installation always prints an error. 
This is my code:

```
@main
struct MyApp: App {
    
    init() {
        ParseSwift.initialize(applicationId: "id", clientKey: "key", serverURL: URL(string: "https://myapp.back4app.io")!)
        DispatchQueue.main.async {
            Installation.current?.save { results in
                switch results {
                case .success(_):
                    print("[Novem] Succesufully saved Installation to ParseServer")
                case .failure(let error):
                    print("[Novem] Failed to update installation: \(error)")
                }
            }
        }
    }
    var body: some Scene {
        WindowGroup {
            MainView()
        }
    }
}
```

I get this error every time my app launches

```Failed to update installation: ParseError(code: ParseSwift.ParseError.Code.unknownError, message: "Error decoding parse-server response: Optional(<NSHTTPURLResponse: 0x281592960> { URL: https://myapp.back4app.io/installations } { Status Code: 200, Headers {\n    \"Access-Control-Allow-Origin\" =     (\n        \"*\"\n    );\n    \"Content-Encoding\" =     (\n        gzip\n    );\n    \"Content-Type\" =     (\n        \"application/json; charset=utf-8\"\n    );\n    Date =     (\n        \"Sun, 14 Mar 2021 14:13:47 GMT\"\n    );\n    Etag =     (\n        \"W/\\\"28-P0BYdfTXNbkssuaZhMIRXlMHiBo\\\"\"\n    );\n    Server =     (\n        nginx\n    );\n    \"Transfer-Encoding\" =     (\n        Identity\n    );\n    \"access-control-allow-credentials\" =     (\n        true\n    );\n    \"access-control-allow-headers\" =     (\n        \"DNT, Keep-Alive, User-Agent, X-Requested-With, If-Modified-Since, Cache-Control, Content-Type, X-Application-ID, X-Access-Token, X-Parse-Master-Key, X-Parse-REST-API-Key, X-Parse-Javascript-Key, X-Parse-Application-Id, X-Parse-Client-Version, X-Parse-Session-Token, X-Requested-With, X-Parse-Revocable-Session, X-CSRF-Token, X-Parse-Client-Key\"\n    );\n    \"access-control-allow-methods\" =     (\n        \"GET, POST, PUT, PATCH, DELETE, OPTIONS\"\n    );\n    \"access-control-expose-headers\" =     (\n        \"X-Parse-Job-Status-Id, X-Parse-Push-Status-Id\"\n    );\n    \"x-powered-by\" =     (\n        Express\n    );\n} }) with error: The data couldn’t be read because it is missing.")```

```
struct Installation: ParseInstallation {
    //: These are required for ParseObject
    var objectId: String?
    var createdAt: Date?
    var updatedAt: Date?
    var ACL: ParseACL?

    //: These are required for ParseInstallation
    var installationId: String?
    var deviceType: String?
    var deviceToken: String?
    var badge: Int?
    var timeZone: String?
    var channels: [String]?
    var appName: String?
    var appIdentifier: String?
    var appVersion: String?
    var parseVersion: String?
    var localeIdentifier: String?
}
```

**Comment by cbaker6** [comment]
Can you share your `Installation` ParseObject and any ParseObjects it embeds?

**Comment by novemTeam** [comment]
I thought the Installation needs to happen every time, how do I make sure it only happens a single time? 
I also want to mention that the first time it does the installation, it returns no errors. 
So it's probably because I try to do it every launch.

But where do I put the installation code then? Is this something we need to check ourselves if it already did the installation?

**Comment by cbaker6** [comment]
> I thought the Installation needs to happen every time, how do I make sure it only happens a single time?

You are not required to save the installation. It's created automatically on and saved to your Keychain on `ParseSwift.initialize`: https://github.com/parse-community/Parse-Swift/blob/0b0d775e404dc6affbb017f271d6469d365f8853/Sources/ParseSwift/Parse.swift#L64-L66

There are only time two times I can think of needing to save it to the Cloud: 1) is when you need it for push notifications. In that case, the only save that's needed occurs in the AppDelegate or equivalent:

```swift
func application(_ application: UIApplication,
                     didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        DispatchQueue.main.async {
            guard var currentInstallation = Installation.current else {
                return
            }
            currentInstallation.setDeviceToken(deviceToken)
            currentInstallation.channels = ["global"]
            currentInstallation.save { _ in }
        }
    }
```

2) If you're making changes to fields in Installation or using Installation for something else.

**Comment by cbaker6** [comment]
The error on your update of Installation might be linked to something with how back4app sends dates. I recommend keeping an eye on this post https://community.parseplatform.org/t/error-decoding-parse-server-response/1446/14

**Comment by cbaker6** [comment]
Closing due to the question being answered and inactivity. Feel free to reopen if needed.

**Comment by cbaker6** [comment]
@novemTeam I think the Installation problems you were having should be fixed in version 1.3.0. In this version you no longer need to wrap installation calls in dispatch closures. Let us know if it resolves the issues you had.

You can see the linked PRs and issues for details.

**Comment by ellekadfur** [comment]
```
func application(_ application: UIApplication,
                     didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        DispatchQueue.main.async {
            guard var currentInstallation = Installation.current else {
                return
            }
            currentInstallation.setDeviceToken(deviceToken)
            currentInstallation.channels = ["global"]
            currentInstallation.save { _ in }
        }
    }
```

same with :
```

    var p = Installation()
            p.setDeviceToken(deviceToken)
            p.channels = [uid]
         p.deviceType = "ios"
        p.create{ result in
                switch result {
                case .success(let installation):
                    // Handle success - `installation` is of type `InstallParse`
                    print("Installation created successfully: \(installation)")
                case .failure(let error):
                    // Handle failure - `error` is of type `ParseError`
                    print("Failed to create installation: \(error.localizedDescription)")
                }
            }

```

Why would I need to set objectId for create() and save()? I'm getting error below @cbaker6 

message	String	"Invalid struct: No value associated with key CodingKeys(stringValue: \"objectId\", intValue: nil) (\"objectId\")."
